haas/rest.py

```diff
@@ -141,49 +122,29 @@ def _do_validation(schema, kwargs):
     `kwargs` should be the arguments to the API call pulled from the URL.
 
     If the schema validates, this will return a dictionary of *all* of the
-    arguments to the function, both from the URL and the body. The argument
-    `kwargs` may be distructively updated.
+    arguments to the function, both from the URL and the body.
 
     If the schema does not validate, this will raise an instance of
     `ValidationError`.
     """
 
-    def _validates(value):
-        """Return True if `schema` validates `value`, False otherwise.
-
-        The schema library's `validate` method raises an exception, which is
-        nice for short-circuting when just doing a simple check, but a bit
-        awkward in more complex cases like the below.
-        """
-        try:
-            schema.validate(value)
-        except SchemaError:
-            return False
-        return True
-
-    if schema is None or _validates(kwargs):
-        return kwargs
-
-    # One innocuous reason validation can fail is simply that we
-    # need parameters from the body (in json). Let's get those and
-    # try again.
-    #
-    # Without `force=True`, this will fail unless the
-    # "Content-Type" header is set to "application/json". We'll be
-    # a little more lienient here; we can discuss whether we want
-    # to tighten this up later, but this makes mucking around with
-    # curl less annoying, and is consistent with pre-flask
-    # behavior.
-    body = flask.request.get_json(force=True)
-
     validation_error = ValidationError(
         "The request body %r is not valid for "
-        "this request." % body)
-    for k in body.keys():
-        if k in kwargs:
-            raise validation_error  # Duplicate key in body + url
-        kwargs[k] = body[k]
-    if not _validates(kwargs):
+        "this request." % flask.request.data)
+
+    try:
+        final_kwargs = json.loads(flask.request.data)
```

Comment:
Could we first check to see if the body is empty before running this check? After that, we can make the ValueError fatal?

That way, if the body is empty we can continue with an empty `final_kwargs`, and we can cut down on unexpected behavior.

For example, the user could mistakenly pass invalid JSON as arguments to an API call that has only optional body arguments. The call would then succeed, but not using the arguments the user thought he had passed.